src/NUnitEngine/nunit.engine.core/Services/ExtensionManager.cs

```diff
@@ -107,84 +132,129 @@ public virtual void FindExtensionPoints(params Assembly[] targetAssemblies)
         {
-            // Create the list of possible extension assemblies,
-            // eliminating duplicates, start in the provided directory.
-            FindExtensionAssemblies(_fileSystem.GetDirectory(startDir));
+            // Ignore a call for a directory we have already used
+            if (!_extensionDirectories.Contains(startDir))
+            {
+                _extensionDirectories.Add(startDir);
 
-            // Check each assembly to see if it contains extensions
-            foreach (var candidate in _assemblies)
-                FindExtensionsInAssembly(candidate);
+                log.Info($"FindExtensionAssemblies examining extension directory {startDir}");
+
+                // Create the list of possible extension assemblies,
+                // eliminating duplicates, start in the provided directory.
+                // In this top level directory, we only look at .addins files.
+                ProcessAddinsFiles(_fileSystem.GetDirectory(startDir), false);
+            }
         }
 
-        /// <inheritdoc/>
-        public void FindStandardExtensions(Assembly hostAssembly)
+        /// <summary>
+        /// Find ExtensionAssemblies for a host assembly using
+        /// a built-in algorithm that searches in certain known locations.
+        /// </summary>
+        /// <param name="hostAssembly">An assembly that supports NUnit extensions.</param>
+        public void FindExtensionAssemblies(Assembly hostAssembly)
         {
+            log.Info($"FindExtensionAssemblies called for host {hostAssembly.FullName}");
+
             bool isChocolateyPackage = System.IO.File.Exists(Path.Combine(hostAssembly.Location, "VERIFICATION.txt"));
             string[] extensionPatterns = isChocolateyPackage
                 ? new[] { "nunit-extension-*/**/tools/", "nunit-extension-*/**/tools/*/" }
                 : new[] { "NUnit.Extension.*/**/tools/", "NUnit.Extension.*/**/tools/*/" };
 
-            FindExtensionAssemblies(hostAssembly, extensionPatterns);
 
-            foreach (var candidate in _assemblies)
-                FindExtensionsInAssembly(candidate);
+            IDirectory startDir = _fileSystem.GetDirectory(AssemblyHelper.GetDirectoryName(hostAssembly));
+
+            while (startDir != null)
+            {
+                foreach (var pattern in extensionPatterns)
+                    foreach (var dir in _directoryFinder.GetDirectories(startDir, pattern))
+                        ProcessDirectory(dir, true);
+
+                startDir = startDir.Parent;
+            }
         }
 
-        /// <inheritdoc/>
+        /// <summary>
+        /// Get an ExtensionPoint based on its unique identifying path.
+        /// </summary>
         public IExtensionPoint GetExtensionPoint(string path)
         {
-            return _pathIndex.ContainsKey(path) ? _pathIndex[path] : null;
+            return _extensionPointIndex.ContainsKey(path) ? _extensionPointIndex[path] : null;
```

Comment:
This is a doing a double lookup.
```suggestion
            _extensionPointIndex.TryGetValue(path, out ExtensionPoint ep);
            return ep;
```

Reply:
Good catch! I preferred to keep the null test for clarity even though it's not needed. Of course, clarity is in the eye of the beholder. :-)
```
return _extensionPointIndex.TryGetValue(path, out ExtensionPoint ep) ? ep : null;
``